### ISSUE TITLE
Wrong/not useful error message when plotting incompatible x and y

### ISSUE THREAD

**Issue by juergspaak** [matplotlib/matplotlib#20452]
When plotting plt.plot(np.ones(10), np.ones((10,0)) it raises a ZeroDivisionError, which confused me much.

**Code for reproduction**

```python
import matplotlib.pyplot as plt
import numpy as np

plt.plot(np.ones(10), np.ones((10,0)))
```

This raises the error:
```
ZeroDivisionError: integer division or modulo by zero
```
**Expected outcome**

I think however, it should either raise a ValueError, similar to plotting plt.plot(np.ones(2), np.ones(10)), i.e. something of the form:
ValueError: x and y must have same first dimension, but have shapes (10,) and (10, 0)

An alternative would be not ploting anything (similar to the behaviour of plt.plot(np.nan, np.nan) ) and potentially raise a warning.

**Matplotlib version**
Matplotlib version 3.3.2
module://ipykernel.pylab.backend_inline

Python version 3.8.5

[@tacaswell edited to add markup]

**Comment by tacaswell** [comment]
Picking out a bit more of the traceback we have

```pytb

~/source/p/matplotlib/matplotlib/lib/matplotlib/axes/_base.py in <genexpr>(.0)
    526 else:
    527     labels = [label] * n_datasets
--> 529 result = (make_artist(x[:, j % ncx], y[:, j % ncy], kw,
    530                       {**kwargs, 'label': label})
    531           for j, label in enumerate(labels))
    533 if return_kwargs:
    534     return list(result)

ZeroDivisionError: integer division or modulo by zero

```

and the error is coming from `0%0`.  We need to check if ncx or ncy is 0 someplace above this, but it is not clear to me if it is better to raise on plot nothing.   I think plotting nothing is more consistent with the way the rest of the broadcasting works?

**Comment by jklymak** [comment]
I don't know 
```
x = np.ones(10)
y = np.ones((10,0))
x+y
``` 
gives: `ValueError: operands could not be broadcast together with shapes (10,) (10,0)`

**Comment by tacaswell** [comment]
However,

```
In [5]: np.ones(10) + np.ones((0, 10))
Out[5]: array([], shape=(0, 10), dtype=float64)
```

and I think that our broadcasting in plot follows the MATLAB column-major rules.

**Comment by jklymak** [comment]
```python
x = np.ones(10)
y = np.ones((0,10))
fig, ax = plt.subplots()
ax.plot(x,y)
```
Gives 
`ValueError: x and y must have same first dimension, but have shapes (10,) and (0, 10)`

**Comment by tacaswell** [comment]
Sorry for not being clear.  What I meant is that in the dimension that numpy does its broadcasting, if you run it to 0 you get back a dimension with 0 on a side.  However for MATLAB reasons, the rules on how we broadcast between the x and y values for multiple datasets is a bit different than the numpy broadcasting progression

```python
ax.plot(np.arange(10), np.ones((10, 3)))  # 3 lines
ax.plot(np.arange(10), np.ones((10, 2)))  # 2 lines
ax.plot(np.arange(10), np.ones((10, 1)))  # 1 lines
# ax.plot(np.arange(10), np.ones((10, 0)))  # 0 lines or expolde?
```

In this progression "do nothing" seems like the right choice.

**Comment by jklymak** [comment]
Thanks,  I guess given that `ax.plot([])` doesn't crash, I agree that `ax.plot(np.arange(10), np.ones((10, 0)))` should not crash as well.

**Comment by tacaswell** [comment]
Labeling as a good first issue as I think it needs a short-circuit return just above the traceback identified above (or maybe where we call that generator from) and a test.  It will require reading and understanding some intermediate level Python, but no API design.

**Comment by jperezg-st** [comment]
Hello, if its okay can I give it a try?

**Comment by tacaswell** [comment]
@jperezg-st Please do!

**Comment by QuLogic** [comment]
Fixed by #20530.
